wagtaildraftail/wagtail_hooks.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import absolute_import, unicode_literals
+
+from draftjs_exporter.constants import BLOCK_TYPES, ENTITY_TYPES, INLINE_STYLES
+from wagtail.wagtailcore import hooks
+
+from .features import BlockFeature, BooleanFeature, EntityFeature, ImageFeature, InlineStyleFeature
+
+
+@hooks.register('register_rich_text_features')
+def register_core_draftail_features(features):
```

Comment:
Maybe more of a question for upstream but if the idea is to have swapable rich text editors, we should ensure the feature name is consistent.

Should there be some feature name constants and/or a list of what is considered the default features (i.e. what an editor should at least implement)?

Reply:
The current list of recognised features is documented at http://docs.wagtail.io/en/v1.12.2/advanced_topics/customisation/page_editing_interface.html#limiting-features-in-a-rich-text-field. It's designed to be a decentralised system, though - both for separation of concerns within Wagtail (the concept of an 'image' or 'document' should be defined entirely within wagtailimages / wagtaildocs respectively; within wagtailcore / wagtailadmin they shouldn't be treated as more 'official' or privileged than any other model a third-party developer might come up with) and to allow it to be extended by third-party modules. As a result, we can't really keep a definitive list of constants within Wagtail core - any new additions will be established by convention.

So, for example, the Wagtail ecosystem might eventually end up with:

* wagtailadmin defining the basic block and style types and providing implementations of them for hallo.js and Draftail
* wagtailimages and wagtaildocs introducing 'image' and 'document-link' features and implementing them for hallo.js and draftail
* a third-party wagtailtinymce module implementing a tinymce-based editor, with implementations of all of the above features
* another third-party module introducing 'table' as a feature, and providing implementations for draftail and tinymce

An editor can implement as few or as many features as it wants, and should ignore any features it doesn't recognise (note: it looks like I need to fix that here :-) ) - but, of course, the more features it supports (and the more it sticks to the existing conventions), the more useful it will be.

Reply:
On a related note, I suspect that when we come to integrate rich text features into the draftjs_exporter process, we'll end up bypassing the constants defined in draftjs_exporter.constants and using the string-based identifiers directly, so that we're not relying on one constants file containing an exhaustive list of all the entity/block/style types we want. I think that's a justifiable change, though - essentially, the rich text features registry does the same job of encapsulating that knowledge (the 'h3' feature maps to the block type 'header-three', etc) and keeping it away from casual tinkering.